uplink/config.go

```diff
@@ -40,10 +43,58 @@ type RSConfig struct {
 // EncryptionConfig is a configuration struct that keeps details about
 // encrypting segments
 type EncryptionConfig struct {
-	Key       string      `help:"root key for encrypting the data"`
-	BlockSize memory.Size `help:"size (in bytes) of encrypted blocks" default:"1KiB"`
-	DataType  int         `help:"Type of encryption to use for content and metadata (1=AES-GCM, 2=SecretBox)" default:"1"`
-	PathType  int         `help:"Type of encryption to use for paths (0=Unencrypted, 1=AES-GCM, 2=SecretBox)" default:"1"`
+	// TODO: WIP#if/v3-1541 read TODO in Key method why the following field cannot
+	// be added
+	// RootKey     *storj.Key  // The root key for encrypting the data and read it from KeyFilepath, see Key method.
+	KeyFilepath string      `help:"the path to the file which contains the root key for encrypting the data"`
+	BlockSize   memory.Size `help:"size (in bytes) of encrypted blocks" default:"1KiB"`
+	DataType    int         `help:"Type of encryption to use for content and metadata (1=AES-GCM, 2=SecretBox)" default:"1"`
+	PathType    int         `help:"Type of encryption to use for paths (0=Unencrypted, 1=AES-GCM, 2=SecretBox)" default:"1"`
+}
+
+// Key returns the root key for encrypting the data which is stored in the path
+// indicated by KeyFilepath field.
+//
+// It returns an error if:
+//
+// * The file doesn't exist.
+// * The file doesn't have exactly 0400 permissions.
+// * There is an I/O error.
+//
+// TODO: WIP#if/v3-1541 refactor cache the key after the firs read. The problem
+// is with pkg/cfgstruct which is inflexible on accepting unexported keys nor
+// several different types as a type of storj.Key, [32]byte, etc.
+func (encCfg *EncryptionConfig) Key() (storj.Key, error) {
+	if encCfg.KeyFilepath == "" {
+		return storj.Key{}, nil
+	}
+
+	f, err := os.Open(encCfg.KeyFilepath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return storj.Key{}, Error.Wrap(fmt.Errorf("not found key file %q", encCfg.KeyFilepath))
+		}
+
+		return storj.Key{}, errs.Wrap(err)
+	}
+
+	fi, err := f.Stat()
+	if err != nil {
+		return storj.Key{}, errs.Wrap(err)
+	}
+
+	if p := fi.Mode().Perm(); p != os.FileMode(0400) {
```

Comment:
`p` -> `perm`

Reply:
Waiting for the var names convention, but for this one it's a very limited scope (one line) that you can see clear what `p` is

Reply:
Question is why you want to keep it `p`? :) Is it more readable or cleaner?

Reply:
> Question is why you want to keep it p

make the lines shorter, it's clear to realize that `p` is the result of `Perm` and make more clear to realize that it's just a var initialized in this due to `:=`

I've never used that many short variables until I started to develop in Go, at the beginning I didn't get it why of such convention, however I got used to and then I've realized that the code was as readable as code with longer names, so it's a matter of what people is use to.

Obviously, they have to use with measure, the longer the scope is the more explicit names are needed, however I also found impossible to have explicit and clear names, which are not needed to go back and see where they are coming, for variables involved in several lines of code.

>  Is it more readable or cleaner?

For the reason commented above and the fact that I found very long lines less readable (i.e. of one instructions, although you break it down in several lines) despite of the use of descriptive names.


---

At the end what I want is not what I wrote, it's to understand some conventions that they are not subjective, to have the next changes aligned on them for having to make less change due to reviews and avoid subjective discussions.

Bear in mind, this is my first more than a few lines PR, and I read the Storj conventions, however I didn't find this topic clear, so I go ahead on what I thought that's clear (not reading my code, reading go code during the last years) to actually have this discussion if somebody mentioned on them and clarify it for understanding what to follow on the future changes.

Reply:
for what it's worth, i think ivan's right that in general, go code prefers brevity (see https://talks.golang.org/2014/names.slide and https://www.reddit.com/r/golang/comments/8wxwgv/why_does_go_encourage_short_variable_names/). we have a preference for slightly longer names in cases where there might be ambiguity (https://github.com/storj/docs/blob/master/code/Style.md#variable-naming) buuut fwiw i think the p choice here is fine. it's limited to just a couple lines of scope

Reply:
😮 I read that Storj conventions part before I've joined, however I didn't remember that it was a reference to an exact (soft limit) number of lines.

Reply:
btw, thanks @jtolds  for the links, I didn't post them, because I though that they were known and mostly I thought that it was another convention and my goal was trying to know which one for bearing in mind for the future.

Reply:
From the beginning I should mark this as a nit. I know that Go prefers brevity and acronyms but most probably it will take a long time for me to use to it. Even if var is declared very close to a place where its used, for me its always additional conversion in my head ;)

Sorry for additional buzz :)

Reply:
Nothing to be sorry, it has clarified me a lot about Storj code conventions regarding this topic.

At the end, the language matters, if yo code in other language which another convention the most of the people expect to find the conventions that are more widely used.

Reply:
I would definitely prefer longer variable names `perm`.

By shortening the line it's not increasing readability. The way to keep things more readable is to move the evaluation to a separate line instead.

```
perm := fi.Mode().Perm(); 
if perm != os.FileMode(0400) {
    return storj.Key{}, Error.New("permissions '%#o' for key file %q are too open", p, encCfg.KeyFilepath)
}
```

Reply:
It looks more a personal opinion.
On the other hand I agree of breaking the lines, however when to do so or not, foe example it looks a common case in the code baes to use

`defer func() { require.NoError(t, f.Close() }()` in one line.

I can perfectly read and understand it, however I don't think that it's very readable.

Reply:
that code can be `defer ctx.Check(f.Close)`. I agree that the previous is not very readable.